Forms: Add Akismet panel to form block

edanzer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Partially addresses: https://github.com/Automattic/jetpack-roadmap/issues/2272

Earlier feedback addressed, and summarized [here](https://github.com/Automattic/jetpack/pull/41826#issuecomment-2683291164).

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* **New Akismet panel**. Adds an Akismet panel to the Jetpack forms block. 
   - Akismet not installed: Shows 'marketing' message and plugin install button.
   - Akismet installed but not active: Shows a plugin activation button.
   - Akismet plugin active: 
      - If valid Akismet key is present, shows buttons to view spam and stats
      - If no valid Akismet key, shows prompt and button to add key
* **Shared PluginIntegrationPanel component.** This PR creates a shared PluginIntegrationPanel component, which the Akismet panel uses. Eventually, I'd like to update the Jetpack CRM and Creative Mail panels (if we keep them) to use this component. Those panels each have four complex files with a lot of duplication. [See this issue for additional details](https://github.com/Automattic/jetpack-roadmap/issues/2254?issue=Automattic%7Cjetpack-roadmap%7C2304). 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
peKye1-1mV-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
Yes.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Ensure Akismet is not installed on your test site.
* Add a form block to page.
* Confirm Spam Protection panel now shows on sidebar.
* If Akismet is not installed, confirm there is prompt to install, plus install button (first screenshot below)
* Click install button, and confirm that Akismet is installed and activated. You should see a progress spinner while the plugin installs. 
* Now go to your plugin page and deactivate Akismet - but leave it installed.
* Return to your form page, or refresh the page. You should now see a prompt to activate the plugin and an activation button. (second screenshot below)
* Click the activate button and confirm the plugin is activated.
* **When Akismet plugin is installed**
   - If you do not have an active Akismet key, confirm you see a prompt and button to add it. Confirm link/button go to right place. 
   - If you have an Akismet key, confirm you see buttons to view spam and stats, and confirm buttons go to right place. 
   
 
## Screenshots:

**Akismet not installed:**
 <img width="1310" alt="install-akismet" src="https://github.com/user-attachments/assets/6a1355ad-e720-49ad-99d1-50e75b44551a" />

**Akismet installed but not active:**
<img width="1318" alt="activate-akismet" src="https://github.com/user-attachments/assets/45cb1aa2-7149-4682-a79b-7efb9a7883b3" />

**Akismet active but no key:**
<img width="1342" alt="akismet-panel-without-key" src="https://github.com/user-attachments/assets/29bd7611-3ba5-470c-bcbe-8d3d70c08200" />

**Akismet active with key:**
<img width="1323" alt="akismet-panel-with-key" src="https://github.com/user-attachments/assets/a006f43c-e9fd-4ce9-9ffb-b52b3aab10db" />

edanzer committed:
Add Akismet panel to forms block

edanzer committed:
changelog

edanzer committed:
Update tracking code to single event

jeherve reviewed (commented):
> New shared component for integrations

Thank you for doing that! As we're starting to have more and more integrations, it makes sense to keep things clean with extracting things that way.

I must say though, that this block sidebar is starting to be pretty long. I wonder if adding one more panel will just not be visible, since there is already so much information in there. 
Would it be an option to have a single "integration" panel that would group different options?

edanzer committed:
Move all plugin panels inside permissions check

edanzer commented on projects/packages/forms/src/blocks/contact-form/edit.js:
```diff
@@ -236,16 +237,22 @@ function JetpackContactFormEdit( { name, attributes, setAttributes, clientId, cl
+									<PanelBody title={ __( 'Creative Mail', 'jetpack-forms' ) } initialOpen={ false }>
+										<NewsletterIntegrationSettings />
+									</PanelBody>
+								</>
```
In addition to adding the Akismet panel, I moved the panel for the Creative Mail plugin inside the canUserInstallPlugins check.

edanzer committed:
Refactor PluginIntegrationPanel component

edanzer committed:
Update Akismet panel

edanzer committed:
Ensure CSS classes follow conventions

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+		return {
+			adminUrl: site?.url ? `${ site.url }/wp-admin/` : null,
+		};
+	}, [] );
```
For getting the admin url, is this approach reliable, or would it be better to get the url via PHP and pass it to our JS via wp_localize_script()?

simison replied:
There's a helper hook somewhere, let's use that.

Users can have sub-folder installs of WP, multi-site installs, etc 

Edit, looks to be `import { getSiteAdminUrl } from '@automattic/jetpack-components';`

edanzer replied:
I'm looking at adopting this method. For some reason, that method is not actually exported from the @automattic/jetpack-components at the moment, so checking on Slack about best approach.

edanzer replied:
Based on feedback on Slack, I've updated to use the [getAdminUrl](https://github.com/Automattic/jetpack/blob/8e30f68bad9351aafefec7d545fb9ee62d8c84d0/projects/js-packages/script-data/src/utils.ts#L28) method from the script-data package.

edanzer replied:
Follow up: I actually changed this again. We also had to solve the problem of the admin url being different for classic vs modern views. I've fixed that by adding a method to get the right admin url in php. See related [comment](https://github.com/Automattic/jetpack/pull/41826#discussion_r1970834244).

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+	);
+};
+
+export default AkismetPanel;
```
I made and effort to find good content to populate the Akismet panel when Akismet is active. It shows "Your forms are protected" so users see the value, and links to the spam page (in the Jetpack Inbox view).

But other suggestions for good content are welcome.

simison replied:
I played around a bit some alternatives — these could also use Akismet icon:

<img width="306" alt="Screenshot 2025-02-25 at 14 05 06" src="https://github.com/user-attachments/assets/53977811-264d-4dd0-846e-e5a4718da50b" />
<img width="270" alt="Screenshot 2025-02-25 at 14 08 12" src="https://github.com/user-attachments/assets/46376241-cb1c-4151-bf16-72f9c7cc14ed" />
<img width="267" alt="Screenshot 2025-02-25 at 14 08 29" src="https://github.com/user-attachments/assets/1da90ba6-bbce-4934-bb4c-55e48ff6cb42" />

I like the last version best as the main CTA is clearer. Note I added mention of "spam" to clarify how it's protected.

We can then revisit copy/design at meetup with Ilona, and anyway we'll move this all to more central location out of sidebar.

edanzer replied:
Updated to last option!

edanzer committed:
Update method for retriving plugin status

edanzer committed:
Update reference to core data store

edanzer commented on projects/packages/forms/src/blocks/contact-form/components/shared/plugin-integration-panel/index.js:
```diff
@@ -0,0 +1,173 @@
+
+	const { pluginStatus, isLoading } = useSelect(
+		select => {
+			const installedPlugins = select( coreStore ).getPlugins();
```
I'm using WP core data store to determine plugin status. In the files for the Jetpack CRM and Creative Mail plugins, we appear to use a Jetpack endpoint. Example: 

https://github.com/Automattic/jetpack/blob/6deb2860206736cfdf2279be79b132693277cbe0/projects/packages/forms/src/blocks/contact-form/components/jetpack-crm-integration/jetpack-crm-integration-settings.js#L8

What's the best approach here?

simison replied:
Here's something similar that might be reusable:

https://github.com/Automattic/jetpack/blob/6deb2860206736cfdf2279be79b132693277cbe0/projects/plugins/jetpack/_inc/client/my-plan/my-plan-body.jsx#L367-L391

edanzer replied:
If it works, I think I'll use the existing implementation.

edanzer committed:
Move title prop

edanzer commented:
> I must say though, that this block sidebar is starting to be pretty long. ....Would it be an option to have a single "integration" panel that would group different options?

Yes, we're having a lot of conversations around this. Few thoughts:
- We'll be adding additional third party integrations. Adding each of them to the sidebar is not scaleable. Adding them in a single panel would likely also start to feel messy. 
- We're considering adding a modal that pops-up with third party integrations kind of like what you'll see for SureForms here https://github.com/Automattic/jetpack-roadmap/issues/2261
- But... creating the scalable solution for this is a medium-term project that still needs a lot of discussion and design work, and we wanted to get something in place for Akismet more immediately. Hence this PR. 
- As a side note, we're also talking about removing the Creative Mail panel sooner rather than later. That was part of our partnership with NFD, but they are no longer actively building that plugin. p1739391617212609-slack-C052XEUUBL4

talldan commented on projects/packages/forms/src/blocks/contact-form/components/shared/plugin-integration-panel/index.js:
```diff
@@ -0,0 +1,173 @@
+		);
+	};
+
+	const renderContent = () => {
```
The way there are separate `renderButton`/`renderContent` functions, that both have very similar (but slightly different) conditional logic makes it hard to see exactly what the main component will output for each different state (e.g. isLoading, error, isInstalled, isActive). While reviewing, I'm cross-checking the two functions, and I've found it pretty difficult to read.

I feel like it might be clearer if it didn't use the `render` functions and all the code were inlined into the return statement with each status a single conditional ... e.g:

```jsx
return (
		<PanelBody title={ title } initialOpen={ initialOpen }>
			<div className="jetpack-plugin-integration__content" aria-live="polite">
			    { isLoading && (
			        // render spinner and status markup
			    ) }
			    { isError && (
			        // render error icon and text
			    ) }
			    { isUninstalled && (
			       // render description and install button
			    ) }
			    { isInstalled && (
			       // render description and activate button
			    ) }
			    { isActive && (
			       // render `children` (I think?)
			    ) }
			</div>
		</PanelBody>
);
```

While this might mean some repetition of markup, I think it would make it easier to read. The only part that sometimes ends up difficult is getting the boolean logic right so that multiple states don't render at the same time! I think it'd also be ok to do early returns and repeat the PanelBody, or extract the inner part into a separate component `PluginIntegrationPanelStatus` to make that possible.

edanzer replied:
@talldan - I'm probably close to you in my preference for simpler, single returns. I started that way and departed a bit just because there's more complexity and conditional logic than felt ideal in a single return. But I'm just fine going back that direction. Will update shortly.

edanzer replied:
@talldan - I've inlined code in the return statement. Rather than inline the button logic, which adds duplication of some complexity, I've put the button in a separate component. Let me know if the logic is easier to follow now.

talldan replied:
Thanks, it's very readable now, I like the solution you settled upon 👍

ntsekouras commented on projects/packages/forms/src/blocks/contact-form/components/shared/plugin-integration-panel/index.js:
```diff
@@ -0,0 +1,173 @@
+import { installAndActivatePlugin, activatePlugin } from '../../../util/plugin-management';
+import './styles.css';
+
+const PluginIntegrationPanel = ( {
```
I don't think we should add a reusable panel component since the plans are clearly to have something different to gather the integrations in a single place. For some reason when I was thought a bit about integrations the `preferences` UI came in mind:

<img width="754" alt="Screenshot 2025-02-25 at 9 50 04 AM" src="https://github.com/user-attachments/assets/d87f1b75-293e-43c1-9afa-75898f83c234" />

I think it's worth pinging the right design folks and get some momentum for this, because we might end up just adding more and more for now in the inspector controls.

simison replied:
+1 reusable panel isn't too important right now; we'll end up removing these panels soon enough.

edanzer replied:
@ntsekouras @simison - Thanks for feedback! 

Updating this response. At first, I thought @ntsekouras was suggesting waiting on adding this until a new 3rd party integration UI was in place. I think the main point is about whether a re-usable panel is worth it. 

On the broader third party panel UI... I've raised moving 3rd party integrations to a modal, but there's no agreement yet. There is likely to be a lengthy amount of debating and P2ing, design (we don't have a designer yet), and implementation before we have a new UI in place. So I wouldn't want to wait on that. As is, this PR allows us to start promoting one of our highest value and most monetizable features in line with the current block interface. 

On the reusable panel... most of the same logic will need to be there, whether in a re-usable panel or a non-reusable one. For me, re-usable is always better :). Especially since we're not sure if/when we'll remove the panels, and there's already a lot of duplicated logic everywhere.

Regardless, if we really prefer non-reusable, I'm happy to adopt that approach. I'll spend like 30 minutes moving the code/logic into a non-reusable, standalone Akismet panel and put that up as a separate PR so we see the difference.

> I think it's worth pinging the right design folks and get some momentum for this, because we might end up just adding more and more for now in the inspector controls.

I'd definitely agree on this. I also like your modal suggestion - though as I said, the modal is still just a proposal with mixed reactions and I'm not sure we'll go that route. Maybe I can P2 it to start a conversation and generate ideas.

edanzer replied:
Update: Just thinking I can spend like 30 minutes moving the code/logic into a non-reusable, standalone Akismet panel and put that up as a separate PR so we see the difference.

edanzer replied:
@ntsekouras @simison I addressed other feedback in this PR first. Once done, I prepped a separate PR without the reusable component, and with all the logic in the AkismetPanel component. 

As I noted above, most of the same logic is still needed, so I don't think there's a big savings in terms of complexity - especially a savings that offsets the potential benefit of having re-usable component for other panels we already have in the interface.

I'd still defer, though, if we prefer this approach:

https://github.com/Automattic/jetpack/pull/42043/files

edanzer replied:
Just a note that I've since updated one part of this PR. So I'd need to adopt that latest change in the PR above too. I'll wait to do that until/if we decide to do away with the shared component.

simison replied:
I'm fine merging the re-usable component; I'll let you decide. 👍 It might be good one for evolving later to reusable card component for the UI outside sidebar.

edanzer replied:
Ok, I'll probably roll with this as is. It's already done, and has a few potential/marginal advantages.

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+						target="_blank"
+						rel="noopener noreferrer"
+					>
+						{ __( 'Review Spam', 'jetpack-forms' ) }
```
Lowercase S; let's move to `Sentence case` from `Title Case`

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+
+	return (
+		<PluginIntegrationPanel
+			title={ __( 'Spam Protection', 'jetpack-forms' ) }
```
Sentence case; "Spam protection"

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+			</p>
+			<div className="jetpack-akismet-panel__buttons">
+				{ adminUrl && (
+					<Button
```
Let's opt-in our components into the new `__next40pxDefaultSize` prop — that way they'll look visually good now and in future without extra effort.

https://wordpress.github.io/gutenberg/?path=/docs/components-button--docs

edanzer replied:
Updated!

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+				) }
+				<Button
+					variant="secondary"
+					href="https://akismet.com"
```
https://akismet.com/support/getting-started/using-akismet-with-your-contact-forms/ seems better link here.

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+				{ adminUrl && (
+					<Button
+						variant="secondary"
+						href={ `${ adminUrl }admin.php?page=jetpack-forms#/responses?status=spam` }
```
The URL won't work for folks who are using the classic view, I landed to this:

<img width="1448" alt="Screenshot 2025-02-25 at 13 58 44" src="https://github.com/user-attachments/assets/e62c61f6-5f66-4e1d-824f-6cfcc531c0e3" />

Go to Inbox, switch to classic from "View" and try the link again.

For classic his one works; `/wp-admin/edit.php?post_status=spam&post_type=feedback`

simison replied:
Just not sure if the different URL should be solved here, or classic/inbox could just redirect to each other's URLs?

simison replied:
BTW fine to merge before solving above URL issue; you'd just merge "active" state in a separate PR

edanzer replied:
Good catch. I assumed the inbox view would always be available if linked to directly. Let me see if I can find a quick solution. We don't want any button going to a dead page.

edanzer replied:
I fixed this. I added a method on the PHP side to get the right forms admin url. Seems like something we'll want to be able to get. I then use that and pass the url to our JS.

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/styles.css:
```diff
@@ -0,0 +1,8 @@
+.jetpack-akismet-panel__buttons {
```
You could alternatively use `<HStack spacing="8">` component ([docs](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-hstack--docs)) but I [suggested](https://github.com/Automattic/jetpack/pull/41826/files#r1969643189) to not have two buttons anyway.

edanzer replied:
Since there's just one button now, I've removed the CSS here.

simison commented on projects/packages/forms/src/blocks/contact-form/components/shared/plugin-integration-panel/index.js:
```diff
@@ -0,0 +1,173 @@
+					<div className="jetpack-plugin-integration__panel-content">
+						<div>
+							{ __(
+								"You already have the plugin installed, but it's not activated.",
```
We should mention Akismet here.

edanzer replied:
I've updated this to take in a prop for the plugin name, and output it here. In this case, it'll say "You already have Akismet installed, but it's not activated."

simison commented on projects/packages/forms/src/blocks/contact-form/components/akismet-panel/index.js:
```diff
@@ -0,0 +1,57 @@
+			installText={ __( 'Install Akismet', 'jetpack-forms' ) }
+			activateText={ __( 'Activate Akismet', 'jetpack-forms' ) }
+			description={ __(
+				"Add one-click spam protection for your forms with Akismet. Simply install the plugin and you're set.",
```
Let's link Akismet plugin page here? https://wordpress.org/plugins/akismet/

Then they can check first what we're talking about, and plugin page seems better than Akismet.com since that's what they'll end up installing with CTA.

edanzer replied:
Added!

edanzer committed:
Update AkismetPanel based on PR feedback

edanzer committed:
Update method for getting site url

edanzer commented on projects/packages/forms/package.json:
```diff
@@ -33,6 +33,7 @@
 	"dependencies": {
 		"@automattic/jetpack-analytics": "workspace:*",
 		"@automattic/jetpack-components": "workspace:*",
+		"@automattic/jetpack-script-data": "workspace:*",
```
This is added, based on feedback, so we can use the [getAdminUrl](https://github.com/Automattic/jetpack/blob/8e30f68bad9351aafefec7d545fb9ee62d8c84d0/projects/js-packages/script-data/src/utils.ts#L28) method.

edanzer committed:
Remove css and move file

edanzer committed:
Simplify plugin integration panel (#42041)

edanzer commented:
UPDATE: I've made the following changes based on feedback:
- @talldan: Refactored PluginIntegrationPanel to have a single return statement
- @simison: 
   - Update titles and button labels to sentence case.
   - Added/updated Akismet links both when not installed and when active
   - Opted in to new __next40pxDefaultSize prop
   - Removed Learn More button when active - it is just the 'View spam' button now. 
   - Adopted getAdminUrl from script-data package based on Slack discussions
- @ntsekouras @simison - As an experiment, created [separate PR](https://github.com/Automattic/jetpack/pull/42043/files) without the shared panel component. 
- Moved AkistmetPanel out of a separate folder. We removed some CSS, which meant separate CSS wasn't needed, which meant there's just a single JS file for the component. 

Still to do:
- Fix the link to view spam with both classic and inbox views.

edanzer committed:
Fix url for classic and modern views

edanzer committed:
Remove unneeded npm dependency

simison commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-akismet-panel.js:
```diff
@@ -0,0 +1,63 @@
+		>
+			<p>
+				{ createInterpolateElement(
+					__( 'Your forms are protected from spam with <a>Akismet</a>!', 'jetpack-forms' ),
```
[Commented elsewhere](https://github.com/Automattic/jetpack/pull/42043#pullrequestreview-2643863679) earlier so copying it here too:

> I'm not entirely sure but we might need to also add "Is configured" check with API keys; I saw some code around which help check if Akismet is configured but I don't remember if the API key is required for Akismet to work.

That's fine as a follow-up tho.

edanzer replied:
I've researched a bit and added a conditional interface that either prompts a user to to add an API key if needed, or else shows buttons to view spam and stats.

edanzer committed:
Ensure panel shows on atomic

edanzer committed:
Fix issue from rebase

edanzer committed:
Conditional display based on Akismet key

edanzer committed:
Add lint exception for HStack

edanzer committed:
Fix Akismet references

jeherve commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-akismet-panel.js:
```diff
@@ -0,0 +1,102 @@
+				),
+				{
+					a: (
+						<a
```
Instead of a link, how about using `ExternalLink` to clearly indicate links that leave the site?

edanzer replied:
Yep, makes sense. Updated!

jeherve commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-akismet-panel.js:
```diff
@@ -0,0 +1,102 @@
+							__( 'Your forms are protected from spam with <a>Akismet</a>!', 'jetpack-forms' ),
+							{
+								a: (
+									<a
```
Same here, we could use `ExternalLink`.

edanzer replied:
Also updated!

jeherve commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-akismet-panel.js:
```diff
@@ -0,0 +1,102 @@
+							{
+								a: (
+									<a
+										href="https://akismet.com/support/getting-started/using-akismet-with-your-contact-forms/"
```
Maybe we can use a Jetpack redirect here, so we can track that link? See `getRedirectUrl` for other uses in the repo.

PCYsg-pY7-p2

edanzer replied:
Sure, happy to use Jetpack redirects. Updated.

edanzer committed:
Use ExternalLink and jetpack redirects
